openxr/src/passthrough.rs

```diff
@@ -0,0 +1,226 @@
+//! The [Passthrough feature](https://support.oculus.com/articles/in-vr-experiences/oculus-features/what-is-passthrough/)
+//! allows you to step outside your view in VR to see a real-time view of your surroundings.
+//!
+//! This feature is exclusive to the Oculus Quest 2 as of March 2022.
+//!
+//! More details about passthrough can be found in the [Oculus Native SDK documentation](https://developer.oculus.com/documentation/native/android/mobile-passthrough/)
+//! as well as in the [OpenXR specification](https://www.khronos.org/registry/OpenXR/specs/1.0/html/xrspec.html#XrPassthroughFB)
+//!
+//! Not all features are currently implemented. Execution control features are implemented, style-oriented features are not.
+use crate::{
+    cvt, raw, session, sys, PassthroughFlagsFB, PassthroughLayerPurposeFB, Result, Session,
+    SessionInner,
+};
+use std::ptr;
+use std::sync::Arc;
+use sys::PassthroughLayerFB;
+
+/// A [passthrough feature].
+///
+/// Requires [`XR_FB_passthrough`] to be enabled.
+///
+/// See the [`PassthroughFB struct`].
+///
+/// [`XR_FB_passthrough`]: https://www.khronos.org/registry/OpenXR/specs/1.0/html/xrspec.html#XR_FB_passthrough
+/// [`PassthroughFB struct`]: https://docs.rs/openxr-sys/latest/openxr_sys/struct.PassthroughFB.html
+/// [passthrough feature]: https://developer.oculus.com/documentation/native/android/mobile-passthrough/#create-and-start-a-passthrough-feature
+pub struct Passthrough {
+    pub(crate) session: Arc<session::SessionInner>,
+    handle: sys::PassthroughFB,
+}
+
+impl Passthrough {
+    /// [Create](https://www.khronos.org/registry/OpenXR/specs/1.0/man/html/openxr.html#xrCreatePassthroughFB) a [passthrough feature].
+    ///
+    /// # Arguments
+    ///
+    /// * `flags` - the passthrough feature flags. Unless `PassthroughFlagsFB::IS_RUNNING_AT_CREATION` is enabled, `Self::start` must be called.
+    ///
+    /// # Errors
+    ///
+    /// This function return `ERROR_EXTENSION_NOT_PRESENT` if [`XR_FB_passthrough`] is disabled.
```

Comment:
Elsewhere we just panic if the function isn't enabled, so I'd prefer to do the same here. The upstream docs also specify a wide array of other possible errors (which we probably shouldn't try to duplicate here).

Also, while docs on internal-only functions are a fine thing, we probably don't need to be as exhaustive as for public API.

Reply:
Sometimes it does, sometimes it doesn't.
https://github.com/Ralith/openxrs/blob/master/openxr/src/foveation_profile_fb.rs#L45
https://github.com/Ralith/openxrs/blob/master/openxr/src/hand_tracker.rs#L41

I fear that if passthrough is disabled in the headset settings this call would fail. Maybe I'm wrong here. If I'm right I would like to be able to keep the program running, maybe log the error, but most importantly warn the end user with a nice dialog and a panic here would make this a lot harder.

Reply:
If passthrough is unsupported, then attempting to create an instance with it enabled must fail, which means these methods cannot be called because no instance exists. It is the application's responsibility to check for extension support and only enable those that are absolutely required or supported.

Reply:
Sounds good. I changed this a panic and trimmed the doc a bit.